### ISSUE TITLE
Feasible loss simplification?

### ISSUE THREAD

**Issue by florian-boehm** [soumith/ganhacks#33]
Hello, I wonder if the following simplifications lead in practice to the same result as the original loss functions:

- Use inverted labels everywhere such that D(x) = 0 and D(G(z)) = 1 in the optimal case.
- Drop the logarithm and even the "+1" could be dropped since constants do not affect the gradient:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;![image](https://user-images.githubusercontent.com/3031621/35889597-8c251a0c-0b9c-11e8-92ad-f9a2624590fc.png)

- Then one can directly use a gradient descent optimizer to do:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;![image](https://user-images.githubusercontent.com/3031621/35889963-ebc5baba-0b9d-11e8-8dc4-c5c8a8e7e141.png)

- And also the generator loss would look like this and can be minimized for training the generator:

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;![image](https://user-images.githubusercontent.com/3031621/35890061-6075cd8c-0b9e-11e8-9652-18c85f4fc24d.png)

- If we assume that the discriminator is good enough and will produce an output close to 1 for D(G(z)) then vanishing gradients should be no problem, correct?

Thank you very much for your help!

Florian

**Comment by ljuvela** [comment]
That starts to look a lot like Wassertein GAN (see e.g https://arxiv.org/abs/1704.00028). They also propose additional loss terms to limit the gradient magnitudes in D.

**Comment by florian-boehm** [comment]
Thank you for pointing this paper out to me. If I have understood it correctly, this one point is worth mentioning:

In case of WGAN the activation function in the last layer of the discriminator should be linear and because the output can not be interpreted as probability anymore the discriminator is then called critic.
